### PR TITLE
feature(xjx): More tests for parallel module, add auto recover option.

### PR DESCRIPTION
**Pull request by sailxjx**
Add some testcases:
1. Throw an exception in the parent process in the child process.
2. When a process exits early, the processes that are still sending messages are not affected.

Add an option `auto_recover` to prevent uncaught exceptions in main function.

### CODE REVIEW

File: ding/framework/tests/test_parallel.py [hunk 1]
```diff
@@ -58,3 +58,124 @@ def star_parallel_main():
...
+        if router._retries == 0:
+            for _ in range(3):
+                router.emit("greeting_0", "")
+                time.sleep(0.1)
+            raise Exception("P1 Error")
+        elif router._retries == 1:
+            router.emit("greeting_0", "recovered_p1")
+            time.sleep(0.1)
+        else:
+            raise Exception("Errored too many times")
```
Comment (zjowowen) [inline ding/framework/tests/test_parallel.py:143]: Should "Too many errors" or "Failed too many times" be better?

Comment (zjowowen) [inline ding/framework/tests/test_parallel.py:143]: Should "Too many errors" or "Failed too many times" be better?

**Comment by zjowowen** [comment]
I think it's a very good and neat work to integrate the restart method into the distributive framework!

**Comment by zjowowen** [comment]
I think it's a very good and neat work to integrate the restart method into the distributive framework!

File: ding/framework/tests/test_parallel.py (hunk 1 above)
Comment (sailxjx) [reply ding/framework/tests/test_parallel.py]: Fixed

**Pull request merged**